New component: ha-expansion-panel

Misiu opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Breaking change

<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

## Proposed change

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

![bc6ec0d6b5943b004415fb8d7cf321c0](https://user-images.githubusercontent.com/1741838/94664051-b966a400-030a-11eb-9b2e-48f8ed9d372e.gif)


This adds a new component `ha-expansion-panel`. I want to use it with the time condition editor.

Supported parameters:

- expanded
- outlined

## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #6726
- This PR is related to issue: https://github.com/home-assistant/frontend/pull/6786, https://github.com/home-assistant/architecture/issues/356
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

Misiu committed:
init

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,135 @@
+        @click=${this._togglePanel}
+      >
+        <div class="summary-content">
+          <p class="summaty-title">
```
```suggestion
          <p class="summary-title">
```

zsarnett replied:
Also, lets not use p here since we are overwriting it anyway. Just use span

Misiu replied:
fixed

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,135 @@
+          </p>
+        </div>
+        <div class="summary-icon ${this.expanded ? "expanded" : ""}">
+          <svg viewBox="0 0 24 24">
+            <path d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"></path>
+          </svg>
```
should use mdi icon with ha-svg-icon

Misiu replied:
done

bramkragten commented:
You should probably use a `<ha-card>` as basis.

Misiu commented:
@bramkragten so I should extend `ha-card`?
I can try.

bramkragten commented:
> @bramkragten so I should extend `ha-card`?
> I can try.

No, not extend, just use it in your render.

zsarnett commented:
Should we use HA-card?

Or should this be the guts of the functionality and can be outside of a `<ha-card>` or can be placed with in one

@bramkragten

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  private _togglePanel(): void {
+    this.expanded = !this.expanded;
```
We should fire an event here to notify that the panel expanded or collapsed

Misiu replied:
I fire one of two: `expanded` or `collapsed`. Not sure of the naming convention in HA.

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,135 @@
+      :host {
+        display: block;
+        width: 100%;
+        box-shadow: 0px 2px 1px -1px rgba(0, 0, 0, 0.2),
```
Don't need. I think, personally, we should build this as a base element with the functionality only and let whatever is using it place it in a `ha-card` or not.

Misiu replied:
This is how the current version looks inside the automation editor:
![image](https://user-images.githubusercontent.com/1741838/92569307-be42b580-f280-11ea-9b95-cd62b70c0503.png)

after removing the padding from `card-content` and adding it to `paper-dropdown-menu-light` and `card-menu` this is the result:
![image](https://user-images.githubusercontent.com/1741838/92569522-08c43200-f281-11ea-8a20-8c5783d0481c.png)

My idea is to show the Time part is at least one of two (`after` and `before`) is filled. same with the Days part - if at least one day is selected then show that part.

Not sure if I should alter `ha-card` styles. Any suggestions?

zsarnett replied:
Dont use Ha-card

Misiu replied:
I placed my panel inside ha-card, I'm not using ha-card in my component.

zsarnett replied:
You should do that. Cards inside of cards are bad

zsarnett replied:
I dont think we should add the box shadow right?

Misiu replied:
done

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,135 @@
+      }
+
+      .summary-content {
+        margin: 12px 0;
```
Too much margin. Specifically on the bottom. The content is not close enough to the summary title

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,135 @@
+
+      .summaty-title {
+        margin: 0;
+        line-height: 1.5;
+        letter-spacing: 0.00938em;
+        font-size: 0.9375rem;
+        font-family: "Roboto", "Helvetica", "Arial", sans-serif;
```
```suggestion
        line-height: 1.2;
        text-overflow: ellipsis;
        overflow: hidden;
```

Misiu replied:
I think ellipsis needs `display`, `white-space` an `width` properties, ref: https://stackoverflow.com/a/7993098/965722
I have a problem with `width` because it needs to be calculated.

zsarnett replied:
I pulled your code and editted it. This is needed and I was able to add it.

Or you can just make it a lot and have the element using it do it :)

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,135 @@
+        margin: 12px 0;
+        display: flex;
+        flex-grow: 1;
+        transition: margin 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
```
```suggestion
        transition: margin 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
        overflow: hidden;
```

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,135 @@
+      }
+
+      .summary-icon {
+        width: 1em;
+        height: 1em;
```
We should use a ha-svg-icon and use the css vars to style the width and height. the base 24px is probably good though

Misiu replied:
I removed icon sizing. After replacing svg with ha-svg-icon it looks much better

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,135 @@
+        font-size: 1.5rem;
+        flex-shrink: 0;
+        user-select: none;
+        transform: rotate(0deg);
```
?

Misiu replied:
I removed that. After replacing svg with ha-svg-icon it looks much better.

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,135 @@
+        overflow: hidden;
+        transition: height 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
+        min-height: 0px;
+        transition-duration: 222ms;
```
why are we overwriting the duration here

Misiu replied:
To be honest I don't know. I based my styles on https://material-ui.com/components/accordion/.
Removed.

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,135 @@
+      .summary {
+        display: flex;
+        padding: 0px 16px;
+        min-height: 48px;
```
actually its this that is too large that is making it not close enough to the content

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,135 @@
+        -webkit-tap-highlight-color: transparent;
+      }
+      .summary.expanded {
+        min-height: 64px;
```
and this

zsarnett replied:
The height of the summary title shouldn't change on expanding

Misiu replied:
I tried to follow the material guidelines.
See other components:
https://material.angular.io/components/expansion/examples
https://material-ui.com/components/accordion/

HA is using a lot of [material components](https://github.com/material-components/material-components-web-components), they follow material guidelines, so I guess HA should do the same (for example we should replace `paper-input` with `mwc-textfield`)

zsarnett replied:
Sure. Still seems a bit far from the header in this component. Maybe its the padding on top of the content

Misiu replied:
I've removed header height animation

zsarnett commented:
We should use javascript to set the `max-height` on the element to the `scroll-height` and then transition the max height instead of the height, either using CSS or Javascript and requestAnimationFrame

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,135 @@
+      >
+        <div class="summary-content">
+          <p class="summaty-title">
+            ${this.title}
```
I feel like this would be better as a slot `<slot name="title"></slot>` So that we can insert more than just a title

zsarnett replied:
![image](https://user-images.githubusercontent.com/18730868/92550220-a63b4980-f220-11ea-9927-b23688f653ff.png)

Misiu replied:
done, there are two slots: `title` and `content`.
Feedback is more than welcome 🙂

Misiu committed:
class name and span

Misiu committed:
events

Misiu committed:
use ha-svg-icon, style adjustments

Misiu committed:
remove duration overrite

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,134 @@
+          </span>
+        </div>
+        <div class="summary-icon ${this.expanded ? "expanded" : ""}">
+          <ha-svg-icon path=${mdiChevronDown}></ha-svg-icon>
```
```suggestion
          <ha-svg-icon .path=${mdiChevronDown}></ha-svg-icon>
```

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,134 @@
+  protected render(): TemplateResult {
+    return html`
+      <div
+        class="summary ${this.expanded ? "expanded" : ""}"
```
use `classMap({expanded: this.expaned})`

Misiu replied:
done

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,134 @@
+            ${this.title}
+          </span>
+        </div>
+        <div class="summary-icon ${this.expanded ? "expanded" : ""}">
```
class map

Misiu replied:
done

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,134 @@
+          <ha-svg-icon path=${mdiChevronDown}></ha-svg-icon>
+        </div>
+      </div>
+      <div class="container ${this.expanded ? "expanded" : ""}">
```
class map. 

Maybe a parent div would be better 🤷

Misiu replied:
done. I removed one classMap and used CSS selector

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,134 @@
+        min-height: 48px;
+        transition: min-height 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
+          background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
+        color: inherit;
+        border: 0;
+        cursor: pointer;
+        margin: 0;
+        outline: 0;
+        position: relative;
+        align-items: center;
+        user-select: none;
+        border-radius: 0;
+        vertical-align: middle;
+        -moz-appearance: none;
+        justify-content: center;
+        text-decoration: none;
+        background-color: transparent;
+        -webkit-appearance: none;
+        -webkit-tap-highlight-color: transparent;
```
I think you need to go through the css and not just copy and paste from the element on md

Misiu replied:
done, I've removed unneeded CSS, but I'm not an expert on that and I only tested on Win10+Chrome.

zsarnett commented:
I really like this Component. I already have a branch that is using it 😄

Misiu commented:
@zsarnett that's why I want this to be part of HA :) just need to polish this a bit more.

zsarnett commented:
Any update on this? Would love to be able to get this in so it can start to be used

Misiu commented:
@zsarnett sadly no updates for now. I'll try to move things forward next week.
I'll optimize CSS and address your comments. 
Not sure how to use ha-card in the code.

zsarnett commented:
I wouldn't use ha card

Misiu committed:
use dot notation for svg path

Co-authored-by: Zack Barett <zackbarett@hey.com>

Misiu committed:
classMap

Misiu committed:
simplify css

Misiu committed:
slots + css tweaks

Misiu commented:
> I wouldn't use ha card.

I agree on that 🙂

> But all good. Just making sure it hasn't been forgotten

I pushed some changes today.
I've simplified CSS, now the code is using classMap and I've added slots.
I'll remove Draft, as I think this is functional.
I'd like to optimize the CSS, by using SASS, but I don't see it being used in the repo.

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,116 @@
+
+@customElement("ha-expansion-panel")
+class HaExpansionPanel extends LitElement {
+  @property({ type: String }) title = "";
```
Not used anymore

Misiu replied:
right 🤦‍♂️

Misiu replied:
fixed

zsarnett commented:
Can you post a new Gif of the component and the transitions

Misiu committed:
remove title property

Misiu commented:
> Can you post a new Gif of the component and the transitions

Sure! But nothing really changed with the animation.

![ha-expansion-panel-1](https://user-images.githubusercontent.com/1741838/93775631-f227be80-fc22-11ea-804a-1bbf6d9777a4.gif)

zsarnett commented:
I still think we need to remove the box shadow. And then just have this be the element with no extra styles. Just the functionality. In your card there it should not have a box shadow

Misiu commented:
It looks a bit weird in my opinion because you don't see the sections. Maybe a border or something?
![ha-expansion-panel-2](https://user-images.githubusercontent.com/1741838/93777395-ffde4380-fc24-11ea-90d2-f1aff7a35d8a.gif)

Besides that the `ha-card` has 16px padding, so it looks even weirder:
![image](https://user-images.githubusercontent.com/1741838/93777532-3451ff80-fc25-11ea-9ef3-256979cf3690.png)

I've removed the padding from `ha-card` and added 8px padding to condition type and dots menu:
![image](https://user-images.githubusercontent.com/1741838/93777914-9f033b00-fc25-11ea-9099-e8760dedc9ff.png)

It looks much better! But I'm not an expert and I have no ideas how to make that global 🙂

zsarnett commented:
You can use `<ha-card outlined>`

But I don't think its needed. Maybe just a border around it. But I don't think we need to limit our selves to only using this in certain places. but use this for functionality and do whatever you want when you use it

Misiu committed:
added outlined property

Misiu commented:
> You can use `<ha-card outlined>`
> 
> But I don't think its needed. Maybe just a border around it. But I don't think we need to limit our selves to only using this in certain places. but use this for functionality and do whatever you want when you use it

I agree on that, the component must be reusable, that's the main idea of a component 🙂
I've added `outlined` property (based on `ha-card` [code](https://github.com/home-assistant/frontend/blob/dev/src/components/ha-card.ts#L37-L45)), this way we can control if we want to add a border or not.

zsarnett commented:
I have copied this code into one of my branches and the animation that opens the Expansion Panel is not good. Its basically none existing... 
![ee0fc7b4ad7522b41d6da8c11afc8905](https://user-images.githubusercontent.com/18730868/93908819-b6ddcc00-fcc4-11ea-9cca-0f3c48e2572b.gif)

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,123 @@
+        display: flex;
+        padding: 0px 16px;
+        min-height: 48px;
+        transition: min-height 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
+          background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
```
```suggestion
        transition: min-height 150ms cubic-bezier(0.4, 0, 0.2, 1),
          background-color 150ms cubic-bezier(0.4, 0, 0.2, 1);
```

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,123 @@
+        margin: 12px 0;
+        display: flex;
+        flex-grow: 1;
+        transition: margin 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
```
```suggestion
        transition: margin 150ms cubic-bezier(0.4, 0, 0.2, 1);
```

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,123 @@
+      }
+
+      .summary-icon {
+        transition: transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
```
```suggestion
        transition: transform 150ms cubic-bezier(0.4, 0, 0.2, 1);
```

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,123 @@
+      .container {
+        height: 0;
+        overflow: hidden;
+        transition: height 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
```
```suggestion
        transition: height 300ms cubic-bezier(0.4, 0, 0.2, 1);
```

Misiu replied:
done

Misiu commented:
@zsarnett I think the cause for this is 0:auto height animation, which doesn't work. A workaround is to use JS animation, but I'd like to avoid that.
Ref: https://css-tricks.com/using-css-transitions-auto-dimensions/

zsarnett commented:
Yea I agree. We need to find a solution for this though. 😃

Misiu committed:
animate max-height instead of height

Misiu commented:
> Yea I agree. We need to find a solution for this though. 😃

I found a solution, but I'm not totally proud of it. I'm animating `max-height`. The downside is that I must set a fixed `max-height` for the container, currently, it is 1000px.
Another option is the JS animation. I'm not a CSS guru like you guys, so is someone proposes a better solution I'll be more than happy to change this. Maybe something like this: https://www.w3schools.com/howto/tryit.asp?filename=tryhow_js_accordion_symbol ??

I really like your usage of this component 🎉, but I think it will look even better if I remove the summary (title) height animation. WDYT?

zsarnett commented:
Yea I researched the animation some time ago and only thing I could find was the Max Height with a set value or Javascript. The Javascript implementation actually isn't terrible but the CSS only also doesn't pose that much of a problem. We can always increase the max height later it just comes down to if we want to ever change the component to accommodate anything larger than 1000px.

zsarnett commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,122 @@
+
+      .container {
+        overflow: hidden;
+        transition: max-height 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
```
```suggestion
        transition: max-height 150ms cubic-bezier(0.4, 0, 0.2, 1);
```

no need for the 0ms delay on the transitions

Misiu replied:
done

Misiu commented:
@zsarnett and what about header animation? Remove it? This doesn't look good in entities editor (your example)

zsarnett commented:
Personally, I think it should be removed but if it is a part of the MD spec then maybe we should keep it /shrug

Misiu committed:
remove header height animation, simplify css

Misiu commented:
@zsarnett I've removed header height animation. I didn't like it either.
I've simplified the CSS.
Please pull the changes into your branch and give this a try.
With your approval, we're ready to go 🙂

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,113 @@
+        </div>
+      </div>
+      <div class="container ${classMap({ expanded: this.expanded })}">
+        <slot name="content"></slot>
```
```suggestion
        <slot></slot>
```

Misiu replied:
done

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,113 @@
+
+  private _toggleContainer(): void {
+    this.expanded = !this.expanded;
+    fireEvent(this, this.expanded ? "expanded" : "collapsed");
```
```suggestion
    fireEvent(this, "expanded-changed", { expanded: this.expanded });
```

Misiu committed:
Update src/components/ha-expansion-panel.ts

single event

Co-authored-by: Bram Kragten <mail@bramkragten.nl>

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,113 @@
+      }
+
+      .container.expanded {
+        max-height: 1000px;
```
This should not be a max-height of 1000px, what if I put 2000px of content in it?

Misiu replied:
I've tried animating height from 0 to auto, but that doesn't work :/
The current solution is using max-height. Another option is to use JS to animate the height, but I won't know the height of content if it will be lazy-loaded.
I'm open for suggestions on this.

bramkragten replied:
Use `display: none; overflow: hidden;` and `display: block;` instead of the height.

Misiu replied:
This doesn't work :/ I want the slide up/down animation.

bramkragten replied:
You should change your transition, it now only works for max-height

bramkragten replied:
Setting it to 1000 will also not animate it correctly right? As it will animate it to 1000 and it will not be height 1000?

bramkragten replied:
Best would probably be to use the height, and remove the height on `transitionend`

Misiu replied:
> Setting it to 1000 will also not animate it correctly right? As it will animate it to 1000 and it will not be height 1000?

RIght, the animation looks different if I use 150 and 5000 as max-height :/

> Best would probably be to use the height, and remove the height on transitionend

But I need to know the height of the content before I start the animation.
I'll search for something over the net

bramkragten replied:
Use `scrollHeight`

Misiu replied:
ok, I got something. Maybe a bit overcomplicated, but it looks and works fine.
I'm not sure how to set the height of an already expanded panel. I used `firstUpdated`. Had to use `setTimeout` because without it height was not accurate (for example, scrollHeight returned 70 instead of 120)

Misiu replied:
this is fixed with `transitionend` event - I'm setting `height: auto`

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,113 @@
+        <div class="summary-content">
+          <slot name="title"></slot>
+        </div>
+        <div class="summary-icon ${classMap({ expanded: this.expanded })}">
+          <ha-svg-icon .path=${mdiChevronDown}></ha-svg-icon>
+        </div>
```
Drop the div

Misiu replied:
done

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,113 @@
+        overflow: hidden;
+      }
+
+      ::slotted([slot="title"]) {
+        display: flex;
+      }
```
Why is this?

Misiu replied:
I've added a couple of spans to that slot, but I think styles should be applied to the slot, not in the component. I'll remove this.

Misiu replied:
done

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,113 @@
+        display: flex;
+        padding: 0px 16px;
+        min-height: 48px;
+        position: relative;
```
Why position relative?

Misiu replied:
Not sure why :) but not needed. I'll remove it.

Misiu replied:
done

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,113 @@
+  protected render(): TemplateResult {
+    return html`
+      <div class="summary" @click=${this._toggleContainer}>
+        <div class="summary-content">
```
I don't think this div is needed? We can move the styles to the parent?

Misiu replied:
done

Misiu committed:
fix after event rename

Misiu committed:
remove slot name

cadavre commented:
I'd call it `ha-accordion`.

Misiu commented:
> I'd call it `ha-accordion`.

It isn't accordion because you only have one content slot inside, so it's an expansion panel, as it allows you to expand/collapse a single panel.

Misiu committed:
remove div around icon + css fixes

Misiu committed:
remove not needed css

Misiu committed:
remove extra div

Misiu committed:
js animation

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,141 @@
+
+    if (this.expanded) {
+      const scrollHeight = container.scrollHeight;
+      const animationDuration = this.getAutoHeightDuration(scrollHeight);
```
Why do we want a variable duration?

Misiu replied:
because I think the animation should be longer if the container if higher. We should have consistent speed.

bramkragten replied:
I think the animation time should be consistent.

Misiu replied:
If the height will be 100px we should animate for 100ms (example time), but if the height will be 1000px the animation should be 1s, so the speed will be consistent.
This is how Material UI is doing, but I don't have a strong opinion about that, so I can set it to a fixed value.

bramkragten replied:
Got a link to the material specs?

Misiu replied:
@bramkragten sadly there isn't any. Ref: https://github.com/material-components/material-components-web/issues/5939
Some info found here: https://material.io/design/motion/speed.html#duration

I've based this component on https://material-ui.com/components/accordion/

_Note: Accordions are no longer documented in the Material Design guidelines, but Material-UI will continue to support them. It was formerly known as the "expansion panel"._

bramkragten replied:
material-ui <> material.io

I would keep this as a reference: <https://material.io/design/motion/speed.html#duration>
Meaning you should choose between the medium or large, but not dynamic.

Misiu replied:
So 300ms for expand and 250 for collapse would be OK? Or can we set this to a single value, let's say 300ms. This will be the easiest solution (no calculation at all). 

I personally think the length of transition should depend on content height, but let's make things simpler in first run. We can adjust that later 🙂

Misiu replied:
Ok done. I've set a fixed value of 300ms.

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,141 @@
+  }
+
+  private _toggleContainer(): void {
+    const container: HTMLDivElement = this.shadowRoot!.querySelector(
+      ".container"
+    )! as HTMLDivElement;
```
```suggestion
    const container = this.shadowRoot!.querySelector(
      ".container"
    )! as HTMLDivElement;
```

Misiu replied:
done

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,141 @@
+      ".container"
+    )! as HTMLDivElement;
+
+    if (this.expanded) {
+      setTimeout(() => {
+        const scrollHeight = container.scrollHeight;
+        const animationDuration = this.getAutoHeightDuration(scrollHeight);
+        container.style.height = `${scrollHeight}px`;
+        container.style.transitionDuration = `${animationDuration}ms`;
+      }, 0);
+    }
+  }
```
You don't want to animate it if it is expanded on first updated? Just show it expanded.

Misiu replied:
I'm not animating it here, I'm setting the height, so it is expanded.
I don't know where else I should calculate the scrollHeight.

bramkragten replied:
You should not. When expanded the height should be determined by the content.

Misiu replied:
done

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,141 @@
+    if (this.expanded) {
+      const scrollHeight = container.scrollHeight;
+      const animationDuration = this.getAutoHeightDuration(scrollHeight);
+      container.style.height = `${scrollHeight}px`;
```
You should remove the height when the transition is done.

Misiu replied:
Ok. Is there a helper I can use or should I use:
```
x.addEventListener("animationend", myEndFunction);
```

bramkragten replied:
`@animationend=${this._handleAnimationEnd}`

Misiu replied:
done with `@transitionend`

Misiu committed:
address comments

Misiu committed:
remove type

Misiu committed:
fixed animation duration

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  private _handleTransitionEnd() {
+    const container = this.shadowRoot!.querySelector(
+      ".container"
+    )! as HTMLDivElement;
```
Use a `@query()` decorator for this

Misiu replied:
code looks much cleaner with that :) thx

Misiu replied:
done

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,128 @@
+    )! as HTMLDivElement;
+
+    if (this.expanded) {
+      container.style.height = "auto";
```
You can just remove the inline styles, as the normal css rules will take over when it isn't animating.

Misiu replied:
I need this because I'm setting the height explicit before I start the collapse animation

bramkragten replied:
I understand, but you can just remove it? As the css has `height: auto`?

Misiu replied:
I'm setting explicit height in `_toggleContainer`, so the auto is overridden. This is used to remove explicit height when the expand ends. You suggested using `height: auto` for the expanded container.
I need an explicit height to have a transition.

Misiu replied:
@bramkragten are you ok with this? I need to have a fixed height to show transitions. That's why I'm setting it before I start both transitions (expand and collapse). I'm also setting auto height when the expand transition ends.

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  private _toggleContainer(): void {
+    const container = this.shadowRoot!.querySelector(
+      ".container"
+    )! as HTMLDivElement;
+
+    const scrollHeight = container.scrollHeight;
+    container.style.height = `${scrollHeight}px`;
```
this is only used when we are going to expand, so only do it when it is needed.

bramkragten replied:
this is only used when we are going to expand, so only do it when it is needed.

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,128 @@
+      ".container"
+    )! as HTMLDivElement;
+
+    const scrollHeight = container.scrollHeight;
+    container.style.height = `${scrollHeight}px`;
```
Is this also needed for the closing transition?

Misiu replied:
yes, because when expanded I have `height: auto`. I need to set explicit height before I start the transition.

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,128 @@
+    if (!this.expanded) {
+      this.expanded = true;
+    } else {
+      setTimeout(() => {
```
Is this so it can transition from the current height to 0? Transition from `auto` to `0` doesn't work probably?

Misiu replied:
yes, that's why I'm setting the fixed height and in setTimeout I'm setting the height to 0.
Is there a way to remove setTimeout?

Misiu committed:
use @query for container

bramkragten commented on src/components/ha-expansion-panel.ts:
```diff
@@ -0,0 +1,123 @@
+
+  private _handleTransitionEnd() {
+    if (this.expanded) {
+      this._container.style.height = "auto";
```
This works right?
```suggestion
      this._container.style removeProperty("height");
```

Misiu replied:
yes, this works perfectly. I've simplified the logic a bit thanks to this.
The last thing I'd like to remove is `setTimeout`, but collapse doesn't work without this.

Misiu committed:
simplify transition logic

bramkragten commented:
We can merge this as soon as we have implemented it somewhere.

Misiu commented:
> We can merge this as soon as we have implemented it somewhere.

I want to implement this in the time condition editor.
My idea is to split time and date into two panels and expand them only if something is filled.
So I'll show the time part if `after` or `before` is filled, same for the weekday section. I want to show it only if the user has changed something (selected a day).
I can create a PR to show what I have in mind.

Also, @zsarnett had usage for this component.

donkawechico commented:
This is awesome! I also have use for it -- I want to make sequence steps collapsible and show their `alias` property as a short-hand description.

balloob commented:
Why are we collapsing parts of a condition instead of the whole condition?

And when we collapse a condition, we should make sure that a summary of that condition is put in the name.

Ie "Time trigger at 10:01"

Once we have such summaries, we can just automatically collapse everything, making it easier to see what's going on in an automation.
